katcp/test/test_server.py

```diff
@@ -333,6 +333,84 @@ def test_on_client_connect(self):
+        start_thread_with_cleanup(self, self.server)
+        s = katcp.Sensor.float("a-sens")
+        s.set(1234, katcp.Sensor.NOMINAL, 5545.5)
+        self.server.add_sensor(s)
+        self.server._send_message = WaitingMock()
+        self.server.wait_running(timeout=1.0)
+        self.assertTrue(self.server.running())
+
+        # We need to pass in the same client_conn for all requests
+        # since strategies are bound to specific connections.
+        client = WaitingMock()
+        self.server._strategies[client] = {}
+
+        req = mock_req("sensor-sampling", "a-sens", "event", client_conn=client)
+        self.server.request_sensor_sampling(req, req.msg).result(timeout=1)
+        inf = req.client_connection.inform
+        inf.assert_wait_call_count(count=1)
+        (inf_msg,) = inf.call_args[0]
+        self._assert_msgs_equal(
+            [inf_msg], (r"#sensor-status 1234.000000 1 a-sens nominal 5545.5",)
+        )
+
+        req = mock_req("sensor-sampling", "a-sens", "period", 1.5, client_conn=client)
+        self.server.request_sensor_sampling(req, req.msg).result()
+        req = mock_req("sensor-sampling", "a-sens", client_conn=client)
+        reply = self.server.request_sensor_sampling(req, req.msg).result()
+        self._assert_msgs_equal([reply], ["!sensor-sampling ok a-sens period 1.5"])
+
+        req = mock_req(
+            "sensor-sampling", "a-sens", "differential", 1.5, client_conn=client
+        )
+        reply = self.server.request_sensor_sampling(req, req.msg).result()
+        self._assert_msgs_equal([reply], ["!sensor-sampling ok a-sens differential 1.5"])
+
+        req = mock_req(
+            "sensor-sampling", "a-sens", "event-rate", 1.5, 2.5, client_conn=client
+        )
+        reply = self.server.request_sensor_sampling(req, req.msg).result()
+        self._assert_msgs_equal(
+            [reply], ["!sensor-sampling ok a-sens event-rate 1.5 2.5"]
+        )
+
+        req = mock_req(
+            "sensor-sampling",
+            "a-sens",
+            "differential-rate",
+            1.5,
+            2.5,
+            3.5,
+            client_conn=client
+        )
+        reply = self.server.request_sensor_sampling(req, req.msg).result()
+        self._assert_msgs_equal(
+            [reply], ["!sensor-sampling ok a-sens differential-rate 1.5 2.5 3.5"]
+        )
+
+    def test_bulk_sensor_sampling_fails_if_not_supported(self):
+        if not self.server.PROTOCOL_INFO.bulk_set_sensor_sampling:
+            start_thread_with_cleanup(self, self.server)
+            self.server._send_message = WaitingMock()
+            self.server.wait_running(timeout=1.0)
+            self.assertTrue(self.server.running())
+
+            req = mock_req("sensor-sampling", "an.int,a.float", "period", 1.5)
+            with self.assertRaises(FailReply):
+                self.server.request_sensor_sampling(req, req.msg).result()
+
+    def test_querying_multiple_sensor_sampling_strategies_fails(self):
```

Comment:
This test should maybe be moved to test_DeviceServer51 since that has the server with the bulk flag set. It think this test here would raise fail reply for the flag not being set as opposed to the contents of the request which is what we want.

Reply:
Good point.  I think we should delete this one, since we have the same test already in the v51  class:  `test_bulk_sensor_sampling_fails_for_no_strategy`.